queryable-spec.ts

```diff
@@ -0,0 +1,141 @@
+  bindings(entries: [RDF.Variable, RDF.Term][]): Bindings;
+  merge(bindings: Bindings[]): Bindings;
+}
+
+/*
+ * Base interfaces to represent query results. These interfaces are generic
+ * with respect to the types of query metadata objects. These can be extended
+ * by implementors.
+ */
+
+interface BaseQueryResultMetadata<OrderItemsType> {
+  cardinality?: number; // Cardinality estimate
+  order?: OrderItemsType[];
+}
+
+interface BaseQueryResult<MetadataType extends BaseQueryResultMetadata<RDF.Variable | termName>> {
+  type: 'bindings' | 'quads' | 'boolean';
+  metadata(opts: Partial<Record<keyof MetadataType, boolean>>): Promise<MetadataType>;
+}
+
+interface QueryResultBindings<MetadataType extends BaseQueryResultMetadata<RDF.Variable>> extends BaseQueryResult<MetadataType> {
+  type: 'bindings';
+  bindings(): Promise<Bindings[]>;
+  stream(): Stream<Bindings>;
+  variables: RDF.Variable[];
+}
+    
+interface QueryResultQuads<MetadataType extends BaseQueryResultMetadata<termName>> extends BaseQueryResult<MetadataType> {
+  type: 'quads';
+  quads(): Promise<RDF.Quad[]>;
+  stream(): Stream<RDF.Quad>;
+}
+
+interface QueryResultBoolean<MetadataType extends BaseQueryResultMetadata<any>> extends BaseQueryResult<MetadataType> {
+  type: 'boolean';
+  value: Promise<boolean>;
+}
+
+type QueryResult<M> = QueryResultBindings<M> | QueryResultQuads<M> | QueryResultBoolean<M>;
+
+/*
+ * Context objects provide a way to pass additional bits information to
+ * implementors, such as but not limited to:
+ * - data sources
+ * - base IRI for IRI resolution
+ * - timestamp for expression evaluation
+ * - query language
+ * - ...
+ */
+
+// SourceType can be anything the query engine defines
+// TODO: we may consider defining some standards, like 'string', RDF.Source, ...
+interface QueryContext<SourceType> {
+  sources: [SourceType, ...SourceType[]];
+  queryTimestamp?: Date; // Required for certain SPARQL operations such as NOW().
+}
+    
+interface QueryStringContext<SourceType> extends QueryContext<SourceType> {
+  queryFormat?: QueryFormat; // defaults to { language: 'SPARQL', version: '1.1', extensions: [] }
+  baseIRI?: string; // Required for parsing SPARQL queries
+};
+
+interface QueryAlgebraContext<SourceType> extends QueryContext<SourceType> {};
+    
+interface QueryFormat {
+  language: string; // Like 'SPARQL'
+  version: string; // Like '1.1'
+  extensions: string[]; // TODO: leave the syntax of these extensions open for now?
+}
+  
+type Algebra = {}; // TODO: define this (possible starting point: https://github.com/joachimvh/SPARQLAlgebra.js)
+
+/* 
+ * Generic query interfaces. These allow engines to return any type of result
+ * object for any type of query, supporting the kind of flexibility required
+ * by engines such as Comunica.
+ */
+
+interface Queryable<SourceType> {
+  query<MetadataType, ContextType extends QueryStringContext<SourceType>>(query: string, context?: ContextType): Promise<QueryResult<MetadataType>>;
```

Comment:
I can see the intention but I don't think the `MetadataType` is useful here. The available metadata will vary based on:
1. The engine – but to implement this interface the engine has to expose the fully generic method; there's no way for the engine to declare what it does support.
2. The form of the query – which is not available here.
3. It could also depend on other factors such as the sources bring used.

Personally I think it would be better (and much more readable) to leave the metadata untyped.

Reply:
I had a feeling that metadata done this way might be somewhat over the top and unnecessary, happy to simplify!